sweep/aggregator.go

```diff
@@ -0,0 +1,349 @@
+)
+
+// UtxoAggregator defines an interface that takes a list of inputs and
+// aggregate them into groups. Each group is used as the inputs to create a
+// sweeping transaction.
+type UtxoAggregator interface {
+	// ClusterInputs takes a list of inputs and groups them into clusters.
+	ClusterInputs(pendingInputs) []inputCluster
+}
+
+type SimpleAggregator struct {
+	// FeeEstimator is used when crafting sweep transactions to estimate
+	// the necessary fee relative to the expected size of the sweep
+	// transaction.
+	FeeEstimator chainfee.Estimator
+
+	// MaxFeeRate is the maximum fee rate allowed within the
+	// SimpleAggregator.
+	MaxFeeRate chainfee.SatPerKWeight
+
+	// FeeRateBucketSize is the default size of fee rate buckets we'll use
+	// when clustering inputs into buckets with similar fee rates within
+	// the SimpleAggregator.
+	//
+	// Given a minimum relay fee rate of 1 sat/vbyte, a fee rate bucket
+	// size of 10 would result in the following fee rate buckets up to the
+	// maximum fee rate:
+	//
+	//   #1: min = 1 sat/vbyte, max (exclusive) = 11 sat/vbyte
+	//   #2: min = 11 sat/vbyte, max (exclusive) = 21 sat/vbyte...
+	FeeRateBucketSize int
+}
+
+// Compile-time constraint to ensure SimpleAggregator implements UtxoAggregator.
+var _ UtxoAggregator = (*SimpleAggregator)(nil)
+
+// NewSimpleUtxoAggregator creates a new instance of a SimpleAggregator.
+func NewSimpleUtxoAggregator(estimator chainfee.Estimator,
+	max chainfee.SatPerKWeight) *SimpleAggregator {
+
+	return &SimpleAggregator{
+		FeeEstimator:      estimator,
+		MaxFeeRate:        max,
+		FeeRateBucketSize: DefaultFeeRateBucketSize,
+	}
+}
+
+// createInputClusters creates a list of input clusters from the set of pending
+// inputs known by the UtxoSweeper. It clusters inputs by
+// 1) Required tx locktime
+// 2) Similar fee rates.
+//
+// TODO(yy): remove this nolint once done refactoring.
+//
+//nolint:revive
+func (s *SimpleAggregator) ClusterInputs(inputs pendingInputs) []inputCluster {
+	// We start by getting the inputs clusters by locktime. Since the
+	// inputs commit to the locktime, they can only be clustered together
+	// if the locktime is equal.
+	lockTimeClusters, nonLockTimeInputs := s.clusterByLockTime(inputs)
+
+	// Cluster the remaining inputs by sweep fee rate.
+	feeClusters := s.clusterBySweepFeeRate(nonLockTimeInputs)
+
+	// Since the inputs that we clustered by fee rate don't commit to a
+	// specific locktime, we can try to merge a locktime cluster with a fee
+	// cluster.
+	return zipClusters(lockTimeClusters, feeClusters)
+}
+
+// clusterByLockTime takes the given set of pending inputs and clusters those
+// with equal locktime together. Each cluster contains a sweep fee rate, which
+// is determined by calculating the average fee rate of all inputs within that
+// cluster. In addition to the created clusters, inputs that did not specify a
+// required lock time are returned.
+func (s *SimpleAggregator) clusterByLockTime(
```

Comment:
what's the actual probability that these have equal locktimes? seems unlikely no? Don't we want to bucket on locktime distance too or is that too dangerous because of needing to aggressively time out htlcs?

Reply:
ah looks like this is just reshuffling logical interfaces.

Reply:
> what's the actual probability that these have equal locktimes?

Happens pretty often when a channel needs to go to chain with a buncha HTLCs which are MPP shards so part of the same logical payment, and if they traverse a similar route, then they'll end up with close or identical CLTV values. 

You need to cluster by lock time to ensure the inputs can be spent in the same transaction. Some of them are also second level inputs, so the lock time is actually only signed in the txn.

Reply:
In theory we can have more intelligent clustering as well, since given a lock time of `T`, and input with a lock time `K` with `K < T` can be bundled together.

Reply:
Yeah that's exactly what I was thinking but we do have to consider how late we are willing to be on sweeping. Conceptually it feels alright if we're at least willing to delay by CLTV∂/2 and then we can ensure that the diameter of the cluster is no larger than that and then sign the transaction with the latest locktime requirement. That said, this is optimization and is certainly delay-able.

Reply:
So this sent me down a rabbit hole where I was trying to convince myself that this was true. Check my reasoning here:

All sighash types commit to nLockTime, so we can't bundle many different second-stage htlc txs unless they have the same locktime because even if we were to change the locktime of the overall tx to satisfy all of the different CLTV constraints in the timeouts, the signatures would be invalidated because of the nLockTime change?